### PR TITLE
Prepare IaC files for a differential scan

### PR DESCRIPTION
**Pull request by Paul-GitGuardian**
As #558 was getting too big, this PR (not functional as-is) will focus on the files part of the IaC diff scan functionality.

### CODE REVIEW

File: tests/unit/core/test_git_shell.py [hunk 1]
```diff
@@ -42,3 +49,123 @@ def test_check_git_dir(tmp_path):
...
+    repo.create_commit()
+
+    # AND a filter function
+    def filter(path, content):
+        return "ignored" not in path
+
+    # AND a list of filepaths
+    filepaths = [first_file_name, first_ignored_file_name]
+
+    # When creating a tar
```
Comment (cyrvan-b) [inline tests/unit/core/test_git_shell.py:159]:
```suggestion
    # WHEN creating a tar
```

**Comment by Paul-GitGuardian** [comment]
@agateau-gg these are mostly changes that you've seen already in #558, with tests added.

File: ggshield/core/git_shell.py [hunk 2]
```diff
@@ -154,3 +172,82 @@ def get_list_commit_SHA(
...
+) -> bytes:
+    """
+    Builds a gzipped archive from a given git reference, and selected filepaths.
+    The filepaths are typically obtained via `get_all_filepaths_from_reference`
+    before being filtered.
+    The archive is returned as raw bytes.
+    :param wd: string path to the git repository. Defaults to current directory
+    :param ref: git reference, like a commit SHA, a relative reference like HEAD~1,
+    or any argument accepted as <ref> by git show <ref>:<filepath>
+    :param filepaths: string paths to selected files
```
Comment (agateau-gg) [inline ggshield/core/git_shell.py:218]: Can you document `acceptation_func`?

File: ggshield/core/git_shell.py [hunk 3]
```diff
@@ -154,3 +172,82 @@ def get_list_commit_SHA(
...
+
+def tar_from_ref_and_filepaths(
+    wd: Optional[str],
+    ref: str,
+    filepaths: List[str],
+    acceptation_func: Optional[Callable[[str, str], bool]] = None,
+) -> bytes:
+    """
+    Builds a gzipped archive from a given git reference, and selected filepaths.
+    The filepaths are typically obtained via `get_all_filepaths_from_reference`
```
Comment (agateau-gg) [inline ggshield/core/git_shell.py:212]:
```suggestion
    The filepaths are typically obtained via `get_filepaths_from_ref`
```

File: ggshield/core/git_shell.py [hunk 4]
```diff
@@ -154,3 +172,82 @@ def get_list_commit_SHA(
...
+        wd = os.getcwd()
+
+    filepaths = git(["ls-files", "-c"], cwd=wd).splitlines()
+    return filepaths
+
+
+def tar_from_ref_and_filepaths(
+    wd: Optional[str],
+    ref: str,
+    filepaths: List[str],
```
Comment (agateau-gg) [inline ggshield/core/git_shell.py:207]: nit: you can make `filepaths` an `Iterable[str]` (or even an `Iterable[Path]`, I would very much like to see more Path in the codebase 😉), this way one can use the return value of a generator as an argument.

File: tests/unit/core/test_git_shell.py [hunk 5]
```diff
@@ -42,3 +49,138 @@ def test_check_git_dir(tmp_path):
     with cd(str(tmp_path)):
         with pytest.raises(UsageError):
             check_git_dir()
+
+
+def test_check_git_ref(tmp_path):
```
Comment (agateau-gg) [inline tests/unit/core/test_git_shell.py:54]: This test should be split at least in two: one test for the not-in-a-repo part and another for the in-a-repo part.

There should be only one of GIVEN/WHEN/THEN in each test.

File: tests/unit/core/test_git_shell.py [hunk 6]
```diff
@@ -42,3 +49,138 @@ def test_check_git_dir(tmp_path):
...
+    second_content = "Second file (included)"
+    second_file.write_text(second_content)
+    repo.add("second.py")
+
+    # WHEN scanning for files, including staged
+    filepaths = get_staged_filepaths(tmp_path)
+
+    # THEN file from first commit is part of filepaths
+    assert "first.py" in filepaths
+    # THEN file from first commit is part of filepaths
```
Comment (agateau-gg) [inline tests/unit/core/test_git_shell.py:135]:
```suggestion
    # AND staged file is part of filepaths
```

File: tests/unit/core/test_git_shell.py [hunk 7]
```diff
@@ -42,3 +49,138 @@ def test_check_git_dir(tmp_path):
...
+    tar_stream = BytesIO(tarbytes)
+    with tarfile.open(fileobj=tar_stream, mode="r:gz") as tar:
+        filenames = tar.getnames()
+
+        # THEN first file in in tar
+        assert first_file_name in filenames
+        # THEN first_ignored file in not in tar
+        assert first_ignored_file_name not in filenames
+        # THEN second file in not in tar
+        assert second_file_name not in filenames
```
Comment (agateau-gg) [inline tests/unit/core/test_git_shell.py:186]: Unless I am mistaken, this can be simplified as `assert filenames == [first_file_name]`, no?

**Review by agateau-gg — CHANGES_REQUESTED** [review]
The git_shell changes looks good overall, only a few nitpicks 👍🏻 .

There is one important aspect to take into account before we can merge this, though: we try to keep the `main` branch in an "always releasable" state. This means the changes to the existing `iac scan` cannot be merged until they are finished.

You have a few options here:

**Option A:**
- Create an integration branch from `main`. something like `main-iac` (it's better if it starts or ends with `main` because then it's easy to apply the CI protections to it).

- make all your PR target main-iac. We review them as they go.

- When your feature is ready to ship, we can file a PR to merge main-iac in main. Since all commits in main-iac have already been reviewed, it won't be necessary to thoroughly review the merge of main-iac into main.

To avoid diverging too much from main, it's a good idea to regularly merge main into main-iac. We have done this in the past.

**Option B:** Revert the changes made to the existing `iac scan` command for now. If necessary, create a hidden command for testing purposes (hidden in the Click-sense: by adding `hidden=True` to the command definition)

File: ggshield/cmd/iac/scan/scan.py [hunk 8]
```diff
@@ -87,9 +56,20 @@ def scan_cmd(
     if directory is None:
         directory = Path().resolve()
     update_context(ctx, exit_zero, minimum_severity, ignore_policies, ignore_paths)
-    result = iac_scan(ctx, directory)
-    scan = IaCScanCollection(id=str(directory), type="path_scan", result=result)
+    result = (
+        iac_scan(ctx, directory)
+        if since is None
+        else iac_diff_scan(ctx, directory, since, staged)
```
Comment (GG-HH) [inline ggshield/cmd/iac/scan/scan.py:62]: If I read correctly, the command is not correct, maybe that it on purpose and will be in another MR ?

File: ggshield/core/git_shell.py [hunk 9]
```diff
@@ -103,21 +108,34 @@ def git_ls(wd: Optional[str] = None) -> List[str]:
...
 
 
+def check_git_ref(wd: Optional[str] = None, ref: Optional[str] = None) -> None:
+    """Check if folder is a git repository and ref is a git reference."""
+    if wd is None:
+        wd = os.getcwd()
+    check_git_dir(wd)
+
+    if not ref or not is_valid_git_commit_ref(ref=ref, wd=wd):
+        raise UsageError("Not a git reference.")
```
Comment (GG-HH) [inline ggshield/core/git_shell.py:136]: Should we add the value (if existing) to help debugging ?

**Review by GG-HH — COMMENTED** [review]
Sounds good overall!

**Comment by Paul-GitGuardian** [comment]
Thank you for the reviews!
Regarding the integration branch: option A is what I did, following your advice on another PR. The name was misleading though, so I renamed it based on your suggestion.

**Review by GG-HH — APPROVED** [review]

**Pull request merged**